### PR TITLE
Add scalafmt

### PR DESCRIPTION
**Pull request by nicl**
Update: a doc with fuller motivation is here: https://docs.google.com/document/d/1S4nNMS-VKWtTiOxCLCJ3bixI1D2wQ19iGjj-fGo8lbM/edit?usp=sharing. Comments there welcome.

**Nb: there will inevitably be conflicts here, so suggestion is that I will rebase and re-run prior to merging.**

Scalastyle is not currently maintained. Scalafmt is now a better option; it has good editor integration and documentation.

A full project (re)format is not ideal, but we really do need standardised formatting and it seems hard to avoid. See e.g. https://github.com/akka/akka/pull/26356 for a similar move.

Other than line-length, scalafmt defaults are used.

**Dev Ex:**

* both Intellij and VSCode (with Metals) will automatically detect and use scalafmt now.
* ~~style warnings are provided when running test~~
* we will configure TC to fail builds if `scalafmtCheckAll` fails
* it will be expected going forward that developers format before pushing (typically through their editor automatically on save)

If you have thoughts on configuration (see https://scalameta.org/scalafmt/docs/configuration.html) add a comment and we can discuss to arrive at a consensus.

### CODE REVIEW

**Comment by johnduffell** [comment]
I'm a big fan of code tooling (formatting/static analysis/etc) especially where it works without anyone even noticing. 👍 

unfortunately the pr is impossible to review properly in its current form due to the size.  as a minimum you might want to highlight the relevant files in the diff, but preferably one PR to fix the formatting in so it can be waved through, and do the tooling changes in a separate pr

**Review by shtukas — APPROVED** [review]

**Comment by shtukas** [comment]
Looks like you ran this before some of the recent changes that went to the main branch recently. Luckily doesn't seem to be that many files. Otherwise +1, what's the worse that could happen 😅

File: .scalafmt.conf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+version=2.5.2
+
+maxColumn = 100
```
Comment (mchv) [inline .scalafmt.conf:3]: @nicl I think the settings are not the best for readability.
Below I am sharing few examples noticed where I think readability degrades

Before
```scala
 lazy val dfpFacebookIaAdUnitCacheJob: DfpFacebookIaAdUnitCacheJob = wire[DfpFacebookIaAdUnitCacheJob]
```

After
```scala
   lazy val dfpFacebookIaAdUnitCacheJob: DfpFacebookIaAdUnitCacheJob =
     wire[DfpFacebookIaAdUnitCacheJob]
```

Before:

```scala
  getMap(articlePage, Edition(request), false).getOrElse("hasShowcaseMainElement", JsBoolean(false)).as[Boolean],
       getMap(articlePage, Edition(request), false).getOrElse("isFront", JsBoolean(false)).as[Boolean],
       getMap(articlePage, Edition(request), false).getOrElse("isLiveBlog", JsBoolean(false)).as[Boolean],
       getMap(articlePage, Edition(request), false).getOrElse("isMinuteArticle", JsBoolean(false)).as[Boolean],
       getMap(articlePage, Edition(request), false).getOrElse("isPaidContent", JsBoolean(false)).as[Boolean],
```

After 
```scala
       getMap(articlePage, Edition(request), false)
         .getOrElse("hasShowcaseMainElement", JsBoolean(false))
         .as[Boolean],
       getMap(articlePage, Edition(request), false)
         .getOrElse("isFront", JsBoolean(false))
         .as[Boolean],
       getMap(articlePage, Edition(request), false)
         .getOrElse("isLiveBlog", JsBoolean(false))
         .as[Boolean],
       getMap(articlePage, Edition(request), false)
         .getOrElse("isMinuteArticle", JsBoolean(false))
         .as[Boolean],
       getMap(articlePage, Edition(request), false)
         .getOrElse("isPaidContent", JsBoolean(false))
         .as[Boolean],
```

Before
```scala
 def comments(key: DiscussionKey): Action[AnyContent] = Action.async { implicit request => getComments(key) }
 def commentsJson(key: DiscussionKey): Action[AnyContent] = Action.async { implicit request => getComments(key) }
 def commentsJsonOptions(key: DiscussionKey): Action[AnyContent] = Action { implicit request => TinyResponse.noContent(Some("GET, OPTIONS")) }
```

After
```scala
   def comments(key: DiscussionKey): Action[AnyContent] =
     Action.async { implicit request => getComments(key) }
   def commentsJson(key: DiscussionKey): Action[AnyContent] =
     Action.async { implicit request => getComments(key) }
   def commentsJsonOptions(key: DiscussionKey): Action[AnyContent] =
     Action { implicit request => TinyResponse.noContent(Some("GET, OPTIONS")) }
```

Comment (nicl) [reply .scalafmt.conf:3]: Thanks @mchv . It's tough because this is obviously so subjective and people are bound to disagree. For what it's worth, I prefer the 'after' versions in your examples!

There's a good list of line length recommendations across languages here:

https://en.wikipedia.org/wiki/Characters_per_line

The majority of which fall below 100. The Scala style guide itself doesn't recommend anything clearly here unfortunately, but does allude to anything over 80 as being sensible for wrapping:

https://docs.scala-lang.org/style/indentation.html#line-wrapping

**Comment by nicl** [comment]
@johnduffell thanks, it's a good point. I'll push again but this time with the changes to add the plugin etc in a separate commit from the actual formatting changes so things should be easier to review.

File: .scalafmt.conf (hunk 1 above)
Comment (shtukas) [reply .scalafmt.conf:3]: Just want to say that I *much* prefer the before version.

Comment (nicl) [reply .scalafmt.conf:3]: For those who would prefer a greater characters per line setting - cc @mchv @shtukas - what would you prefer?

The other thing I'd add is that once you go over about 100 you stop being able to get two code windows open side by side on a standard resolution (or simply, if you have one coding window and then say some docs etc on the other half), which is another thing to consider.

Comment (sndrs) [reply .scalafmt.conf:3]: fwiw, prettier default line length is 80 chars

Comment (johnduffell) [reply .scalafmt.conf:3]: I agree that 100 is not the best use of available space for coding in an IDE on a decent monitor, but I'm also conscious that we're only 2 steps away from spaces vs tabs debate! 😓 However I wanted to point out we need to consider more than one case here.

If you're reading diffs in github or merging branches in intellij, clarity is maintained better with short lines, especially with side by side diff.  Ironically in @mchv comment the shorter lines are more readable because of no need to scroll sideways
image

Also we are discussing the current code being auto reformatted which is never going to be perfect.  Having a shorter line length will encourage people to write their code differently to fit in with that.

At the end of the day everyone is going to have a different figure based of how they read code more often, my personal opinion is to allow more than 100 but encourage short lines where possible.  Code is communication/documentation so I think we should take more account of what people who use the codebase think than recommendations online, because some of them will still be based around the 80 character standard terminal size!

File: build.sbt [hunk 2]
```diff
@@ -5,139 +5,188 @@ import com.typesafe.sbt.web.SbtWeb.autoImport._
...
-    atomRenderer,
-    identityModel,
-    capiAws,
-    okhttp,
-    jsonSchema
-  ) ++ jackson
-).settings(
-    mappings in TestAssets ~= filterAssets
+val sharedSettings = Seq(
+  (test in Test) := (test in Test).dependsOn(scalafmtCheckAll).value
```
Comment (johnduffell) [inline build.sbt:9]: have you tried
`ThisBuild / Test / test := (Test / test).dependsOn(scalafmtCheckAll).value`
Not totally sure whether it wouldn't be overridden at project level, but try it, that might save having to do it individually on each project.

**Comment by adamnfish** [comment]
I've chatted about this a little in other forums.

I agree your thinking generally, but disagree slightly with some of the details. I argue for three pieces of moderation in your approach, but this is your codebase and I wholly support whatever decision you take.

TL;DR; I ask for the following:

* talk to Data Tech first, who have already done this to the (huge) data lake repository
* start small - first apply this to the next few Scala changes that come down the pipeline and see if it really offers a benefit, before diving into doing the whole repo at once up front
* separate the one-off formatting change from the tooling changes to maintain the change so we can meaningfully review the latter

Here are my thoughts in a little more detail.

## I don't think this should be hooked into the sbt test command.

I feel pretty strongly about this one, but I think there's been agreement here so no need for me to go into detail on it in this comment. I understand this PR will be updated to encourage developers to do the formatting from their editor / workflow.

Unfortunately, the nature of this PR means that change will be very hard to spot. Is it possible to do the one-off change as a single PR and *separately* make changes to the repository to maintain this format in the future?

## I agree that removing decision-making from formatting is a great thing

Like Nic, I've had great experiences in other languages where tooling is set up along these lines. I agree it provides real value to teams and departments, so if Frontend are thinking about doing the same here then that makes sense.

I would encourage you to talk to the Data Tech team before you merge this PR. They have been using formatting tools on their (large, Scala) codebase for some time. They'll be able to give you some insights on how well it has worked in our department in the past, without having to go nuclear on the Frontend repository to try it out first.

## Dotcom is a platform team

Ultimately, the dotcom team are responsible for this repository. If there's a strong feeling from that team that this is worthwhile, then we all support you. This is double-edged though, as a platform we all need to be able to work on it and all deserve at last oversight on this process. I applaud your open and transparent process, and the great communication here and in other channels.

Thanks to @nicl and the team!

## My views

I have a couple of concerns with this process, none of which are huge deals.

### Scala is not well-suited to automatic code formatting

Scala's syntax is unusually complex as languages go, the flexibility of approaches it means that flexibility in syntactical choices is required. My experience of Scala formatting has been that it works fine in most cases but in some cases cause real harm to the legibility of code. I think this will be especially true of Frontend, which is a codebase of varying Scala quality, embracing as it does old ideas and new.

I'm not super interested in the details of the linting, like Nic I prefer a linter with no options at all if we're going to choose one. My question is not "can we get the settings right", but "do right settings exist". On this point there are trade-offs both ways, the team's preference to do the work to migrate more than outweighs factors either way.

### I'm not yet convinced this is worth the effort

Nic has pointed me at a recent Scala PR in Frontend, perhaps the one that caused this change to get made. However, from my (brief) research, I don't generally see a lot of work being done on the Scala part of this codebase. This is a huge change that has already taken up a lot of time from the department, with more ahead. It also introduces new setup steps for developers working on Frontend, new rules that are not generally accepted as Scala best practices, and additional tooling that needs to be maintained directly (on the codebase) and indirectly (as part of the builds etc).

You are better placed than me to assess this though! As the team most likely to be making these changes it's up to you (not me) to decide whether this is worth the effort.

However, I'd encourage you to evaluate this not just conceptually "linted code is a good thing", but on the actual results of linting the Frontend codebase in practice. Is the reformatted code *actually* better? This is why I'd prefer to have seen this rolled out slowly. Let's see whether **as we work on the project**, we find that formatting/linting has improved the legibility. If it does then great, we can roll it out. If not, then we haven't nuked the whole repo and done heaps of tooling work for no reason. Small changes are good changes, no big bang releases, measure and iterate etc etc.


All that said, my real view is that you all know best, and are free to make choices that support your work.

Thanks again for your great work here, and on being so forthcoming and helpful throughout the process.

File: .scalafmt.conf (hunk 1 above)
Comment (nicl) [reply .scalafmt.conf]: Cc @mchv @shtukas @tomrf1 and @johnduffell . Thanks for feedback on this. I've set this to 120 for now, which is used in a few Scala projects and 'fixes' the examples above. Depending on how it goes, we might consider reducing this further in future, but 120 feels like a good compromise at the moment that isn't too restrictive but still provides some kind of upper bound.

Comment (gtrufitt) [reply .scalafmt.conf]: I find these to be far better readability, but that's likely because I am used to look at every JS / TS codebase with prettier on. I guess that's the point - when you're used to reading code with a standard formatting style, there's a far lower cognitive load when coming to an unfamiliar codebase (or unfamiliar part of a huge codebase!).

**Review by jfsoul — APPROVED** [review]
Lots of fair points raised here and elsewhere, but the team consensus is that this will be a positive change. I think compromising with a longer line length to start with (120) and dropping it depending on our experience going forwards is a good move 👍 

I doubt there will ever be any tool or configuration that gets the formatting right every time, but my opinion at least from skimming a lot of the files is that the changes are overwhelmingly for the better - but I know that is subjective.

**Comment by nicl** [comment]
Hey all, this is now live and comms sent out. @adamnfish thanks for the feedback. Just to respond to your specific points that were outstanding, the team decided on a full format in the end but we'll keep an ear and eye out for any issues/feedback as we see how it goes. I had a chat with Amy who worked on data tech for a good while. Unfortunately, it looks like they don't run it as part of their CI so that side of the picture isn't there. Lastly, the commits for the tooling code and actual formatting have been separated, which is definitely an improvement for understandability.
